```
/__w/dealii/dealii/source/lac/scalapack.cc:68:3: error: ‘hid_t dealii::{anonymous}::hdf5_type_id(const char*)’ defined but not used [-Werror=unused-function]
   68 |   hdf5_type_id(const char *)
      |   ^~~~~~~~~~~~
/__w/dealii/dealii/source/lac/scalapack.cc:62:3: error: ‘hid_t dealii::{anonymous}::hdf5_type_id(const unsigned int*)’ defined but not used [-Werror=unused-function]
   62 |   hdf5_type_id(const unsigned int *)
      |   ^~~~~~~~~~~~
/__w/dealii/dealii/source/lac/scalapack.cc:56:3: error: ‘hid_t dealii::{anonymous}::hdf5_type_id(const int*)’ defined but not used [-Werror=unused-function]
   56 |   hdf5_type_id(const int *)
      |   ^~~~~~~~~~~~
```

--- a/include/deal.II/fe/fe_values_base.h
+++ b/include/deal.II/fe/fe_values_base.h
@@ -1807,6 +1807,21 @@ private:
   friend class FEValuesViews::Tensor;
 };
 
+namespace internal
+{
+  /**
+   * Compute the row indices of shape functions in internal data tables. If a
+   * component is zero then the value in this vector is
+   * numbers::invalid_unsigned_int.
+   *
+   * @note This utility function is used by FEValues and FEValuesViews and is
+   * here to avoid duplication.
+   */
+  template <int dim, int spacedim>
+  std::vector<unsigned int>
+  make_shape_function_to_row_table(const FiniteElement<dim, spacedim> &fe);
+} // namespace internal
+
 #ifndef DOXYGEN
 
 /*---------------------- Inline functions: FEValuesBase ---------------------*/

--- a/source/dofs/dof_tools.cc
+++ b/source/dofs/dof_tools.cc
@@ -276,7 +276,7 @@ namespace DoFTools
     // triangulation then the output array is index by
     // dof.locally_owned_dofs().index_within_set(indices[i])
     template <int dim, int spacedim>
-    inline void
+    void
     get_block_association(const DoFHandler<dim, spacedim> &dof,
                           std::vector<unsigned char>      &dofs_by_block)
     {

--- a/source/fe/fe_values.cc
+++ b/source/fe/fe_values.cc
@@ -42,38 +42,6 @@ DEAL_II_NAMESPACE_OPEN
 
 namespace internal
 {
-  template <int dim, int spacedim>
-  inline std::vector<unsigned int>
-  make_shape_function_to_row_table(const FiniteElement<dim, spacedim> &fe)
-  {
-    std::vector<unsigned int> shape_function_to_row_table(
-      fe.n_dofs_per_cell() * fe.n_components(), numbers::invalid_unsigned_int);
-    unsigned int row = 0;
-    for (unsigned int i = 0; i < fe.n_dofs_per_cell(); ++i)
-      {
-        // loop over all components that are nonzero for this particular
-        // shape function. if a component is zero then we leave the
-        // value in the table unchanged (at the invalid value)
-        // otherwise it is mapped to the next free entry
-        unsigned int nth_nonzero_component = 0;
-        for (unsigned int c = 0; c < fe.n_components(); ++c)
-          if (fe.get_nonzero_components(i)[c] == true)
-            {
-              shape_function_to_row_table[i * fe.n_components() + c] =
-                row + nth_nonzero_component;
-              ++nth_nonzero_component;
-            }
-        row += fe.n_nonzero_components(i);
-      }
-
-    return shape_function_to_row_table;
-  }
-} // namespace internal
-
-
-
-namespace internal
-{
   namespace FEValuesImplementation
   {
     template <int dim, int spacedim>

--- a/source/fe/fe_values_base.cc
+++ b/source/fe/fe_values_base.cc
@@ -44,7 +44,7 @@ DEAL_II_NAMESPACE_OPEN
 namespace internal
 {
   template <int dim, int spacedim>
-  inline std::vector<unsigned int>
+  std::vector<unsigned int>
   make_shape_function_to_row_table(const FiniteElement<dim, spacedim> &fe)
   {
     std::vector<unsigned int> shape_function_to_row_table(

--- a/source/fe/fe_values_base.inst.in
+++ b/source/fe/fe_values_base.inst.in
@@ -16,6 +16,13 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : SPACE_DIMENSIONS)
   {
 #  if deal_II_dimension <= deal_II_space_dimension
     template class FEValuesBase<deal_II_dimension, deal_II_space_dimension>;
+
+    namespace internal
+    \{
+      template std::vector<unsigned int>
+      make_shape_function_to_row_table(
+        const FiniteElement<deal_II_dimension, deal_II_space_dimension> &);
+    \}
 #  endif
   }
 

--- a/source/fe/fe_values_views.cc
+++ b/source/fe/fe_values_views.cc
@@ -33,42 +33,6 @@
 DEAL_II_NAMESPACE_OPEN
 
 
-namespace internal
-{
-  namespace
-  {
-    template <int dim, int spacedim>
-    std::vector<unsigned int>
-    make_shape_function_to_row_table(const FiniteElement<dim, spacedim> &fe)
-    {
-      std::vector<unsigned int> shape_function_to_row_table(
-        fe.n_dofs_per_cell() * fe.n_components(),
-        numbers::invalid_unsigned_int);
-      unsigned int row = 0;
-      for (unsigned int i = 0; i < fe.n_dofs_per_cell(); ++i)
-        {
-          // loop over all components that are nonzero for this particular
-          // shape function. if a component is zero then we leave the
-          // value in the table unchanged (at the invalid value)
-          // otherwise it is mapped to the next free entry
-          unsigned int nth_nonzero_component = 0;
-          for (unsigned int c = 0; c < fe.n_components(); ++c)
-            if (fe.get_nonzero_components(i)[c] == true)
-              {
-                shape_function_to_row_table[i * fe.n_components() + c] =
-                  row + nth_nonzero_component;
-                ++nth_nonzero_component;
-              }
-          row += fe.n_nonzero_components(i);
-        }
-
-      return shape_function_to_row_table;
-    }
-  } // namespace
-} // namespace internal
-
-
-
 namespace FEValuesViews
 {
   template <int dim, int spacedim>

--- a/source/grid/grid_generator.cc
+++ b/source/grid/grid_generator.cc
@@ -3424,9 +3424,8 @@ namespace GridGenerator
   namespace internal
   {
     // helper function to check if point is in 2d box
-    bool inline point_in_2d_box(const Point<2> &p,
-                                const Point<2> &c,
-                                const double    radius)
+    bool
+    point_in_2d_box(const Point<2> &p, const Point<2> &c, const double radius)
     {
       return (std::abs(p[0] - c[0]) < radius) &&
              (std::abs(p[1] - c[1]) < radius);

--- a/source/grid/grid_tools.cc
+++ b/source/grid/grid_tools.cc
@@ -257,7 +257,7 @@ namespace GridTools
      * of the @p dim space dimensions. Factorized into a function of its own
      * in order to allow parallel execution.
      */
-    inline void
+    void
     laplace_solve(const SparseMatrix<double>      &S,
                   const AffineConstraints<double> &constraints,
                   Vector<double>                  &u)

--- a/source/grid/grid_tools_dof_handlers.cc
+++ b/source/grid/grid_tools_dof_handlers.cc
@@ -2379,7 +2379,7 @@ namespace GridTools
    * is parallel to the unit vector in <direction>
    */
   template <int spacedim>
-  inline bool
+  bool
   orthogonal_equality(const Point<spacedim>     &point1,
                       const Point<spacedim>     &point2,
                       const unsigned int         direction,

--- a/source/grid/tria.cc
+++ b/source/grid/tria.cc
@@ -2861,7 +2861,7 @@ namespace internal
      * Furthermore, the type of the quad is determined.
      */
     template <int dim>
-    inline void
+    void
     build_intersection(
       const std::vector<ReferenceCell<dim>>           &cell_types,
       const ArrayOfArrays                             &cells_to_vertices,
@@ -15744,14 +15744,14 @@ namespace internal
 {
   namespace TriangulationImplementation
   {
-    inline unsigned int
+    unsigned int
     n_cells(const internal::TriangulationImplementation::NumberCache<1> &c)
     {
       return c.n_lines;
     }
 
 
-    inline unsigned int
+    unsigned int
     n_active_cells(
       const internal::TriangulationImplementation::NumberCache<1> &c)
     {
@@ -15759,14 +15759,14 @@ namespace internal
     }
 
 
-    inline unsigned int
+    unsigned int
     n_cells(const internal::TriangulationImplementation::NumberCache<2> &c)
     {
       return c.n_quads;
     }
 
 
-    inline unsigned int
+    unsigned int
     n_active_cells(
       const internal::TriangulationImplementation::NumberCache<2> &c)
     {
@@ -15774,14 +15774,14 @@ namespace internal
     }
 
 
-    inline unsigned int
+    unsigned int
     n_cells(const internal::TriangulationImplementation::NumberCache<3> &c)
     {
       return c.n_hexes;
     }
 
 
-    inline unsigned int
+    unsigned int
     n_active_cells(
       const internal::TriangulationImplementation::NumberCache<3> &c)
     {

--- a/source/lac/scalapack.cc
+++ b/source/lac/scalapack.cc
@@ -29,44 +29,29 @@ DEAL_II_NAMESPACE_OPEN
 
 #ifdef DEAL_II_WITH_HDF5
 
-template <typename number>
-inline hid_t
-hdf5_type_id(const number *)
+namespace
 {
-  DEAL_II_NOT_IMPLEMENTED();
-  // don't know what to put here; it does not matter
-  return -1;
-}
+  template <typename number>
+  hid_t
+  hdf5_type_id(const number *)
+  {
+    DEAL_II_NOT_IMPLEMENTED();
+    // don't know what to put here; it does not matter
+    return -1;
+  }
 
-inline hid_t
-hdf5_type_id(const double *)
-{
-  return H5T_NATIVE_DOUBLE;
-}
+  hid_t
+  hdf5_type_id(const double *)
+  {
+    return H5T_NATIVE_DOUBLE;
+  }
 
-inline hid_t
-hdf5_type_id(const float *)
-{
-  return H5T_NATIVE_FLOAT;
-}
-
-inline hid_t
-hdf5_type_id(const int *)
-{
-  return H5T_NATIVE_INT;
-}
-
-inline hid_t
-hdf5_type_id(const unsigned int *)
-{
-  return H5T_NATIVE_UINT;
-}
-
-inline hid_t
-hdf5_type_id(const char *)
-{
-  return H5T_NATIVE_CHAR;
-}
+  hid_t
+  hdf5_type_id(const float *)
+  {
+    return H5T_NATIVE_FLOAT;
+  }
+} // namespace
 #endif // DEAL_II_WITH_HDF5
 
 

--- a/source/lac/trilinos_sparse_matrix.cc
+++ b/source/lac/trilinos_sparse_matrix.cc
@@ -1957,13 +1957,13 @@ namespace TrilinosWrappers
     namespace SparseMatrixImplementation
     {
       template <typename VectorType>
-      inline void
+      void
       check_vector_map_equality(const Epetra_CrsMatrix &,
                                 const VectorType &,
                                 const VectorType &)
       {}
 
-      inline void
+      void
       check_vector_map_equality(const Epetra_CrsMatrix              &m,
                                 const TrilinosWrappers::MPI::Vector &in,
                                 const TrilinosWrappers::MPI::Vector &out)

--- a/source/multigrid/mg_transfer_internal.cc
+++ b/source/multigrid/mg_transfer_internal.cc
@@ -453,7 +453,7 @@ namespace internal
 
 
     // Transform the ghost indices to local index space for the vector
-    inline void
+    void
     copy_indices_to_mpi_local_numbers(
       const Utilities::MPI::Partitioner          &part,
       const std::vector<types::global_dof_index> &mine,

--- a/source/non_matching/quadrature_generator.cc
+++ b/source/non_matching/quadrature_generator.cc
@@ -219,7 +219,7 @@ namespace NonMatching
        * Return true if the incoming function bounds correspond to a function
        * which is indefinite, i.e., that is not negative or positive definite.
        */
-      inline bool
+      bool
       is_indefinite(const std::pair<double, double> &function_bounds)
       {
         if (function_bounds.first > 0)
@@ -251,7 +251,7 @@ namespace NonMatching
        * Note that the returned value can be negative. This is used to indicate
        * "how far away" a function is from being definite.
        */
-      inline double
+      double
       lower_bound_on_abs(const std::pair<double, double> &function_bounds)
       {
         Assert(function_bounds.first <= function_bounds.second,
@@ -332,7 +332,7 @@ namespace NonMatching
        * one being negative definite. Return false otherwise.
        */
       template <int dim>
-      inline bool
+      bool
       one_positive_one_negative_definite(
         const std::vector<FunctionBounds<dim>> &all_function_bounds)
       {
@@ -1041,7 +1041,7 @@ namespace NonMatching
        * direction and return the left half.
        */
       template <int dim>
-      inline BoundingBox<dim>
+      BoundingBox<dim>
       left_half(const BoundingBox<dim> &box, const unsigned int direction)
       {
         AssertIndexRange(direction, dim);
@@ -1060,7 +1060,7 @@ namespace NonMatching
        * direction and return the right half.
        */
       template <int dim>
-      inline BoundingBox<dim>
+      BoundingBox<dim>
       right_half(const BoundingBox<dim> &box, const unsigned int direction)
       {
         AssertIndexRange(direction, dim);

